Deprecate x.T on tensors of dimension other than 0 or 2

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64181
* __->__ #64180
* #64179

**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

Differential Revision: [D31610611](https://our.internmc.facebook.com/intern/diff/D31610611)

lezcano committed:
Deprecate x.T on tensors of dimension other than 2

[ghstack-poisoned]

rgommers reviewed (commented):
`dim != 2` or `> 2`? I had expected the latter. Using `.T` on 0-D or 1-D tensors is perfectly well-behaved, right?

Lezcano commented:
For `dim = 0` I understand it's the identity, but what does it mean to transpose a vector? To `unsqueeze(-1)`?

rgommers commented:
> but what does it mean to transpose a vector? To `unsqueeze(-1)`?

No, it does what it does now in code (i.e. it's a no-op).

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2171,6 +2171,13 @@ std::vector<Tensor> meshgrid(TensorList tensors) {
+  if (n != 2) {
+    TORCH_WARN_ONCE(
+        "The use of `x.T` on tensors of dimension other than 2 is deprecated and will throw ",
+        "an error in a future release. Consider `x.mT` to transpose batches of matrices or ",
```
It's probably worth saying why it's deprecated, aka because it probably doesn't do what you want when things are batched

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

ezyang commented:
letting @anjali411 do this one too since she's got the PR below this in the stack

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc mruberry rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2175,7 +2175,14 @@ std::vector<Tensor> meshgrid(TensorList tensors) {
+  const auto n = self.dim();
+  if (n != 2 && n != 0) {
+    TORCH_WARN_ONCE(
+        "The use of `x.T` on tensors of dimension other than 2 to reverse their shape is deprecated ",
```
It's unfortunate we don't have a good consistent way of referring to the tensor whose attribute is accessed. Maybe we can just drop the `x` and say "The use of .T on tensors..."?

Lezcano replied:
I prefer the user of `x` as a placeholder, same as in
https://github.com/pytorch/pytorch/blob/dbfd4d0d563a08a349a8105786d0aa3ee2c09823/torch/_tensor_docs.py#L4845

mruberry replied:
Sure

mruberry commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2175,7 +2175,14 @@ std::vector<Tensor> meshgrid(TensorList tensors) {
+    TORCH_WARN_ONCE(
+        "The use of `x.T` on tensors of dimension other than 2 to reverse their shape is deprecated ",
+        "and it will throw an error in a future release. Consider `x.mT` to transpose batches of matrices",
+        "or `x.permute(*torch.arange(x.ndim - 1, -1, -1))` to reverse the dimensions of a tensor."
```
I would if this functionality is so common that we should add a special operator, like `reverse_dims` as a drop-in replacement?

Lezcano replied:
I do not know of any use case of this. I just wrote how to implement the functionality in PyTorch in case anyone wants to use it, but I highly doubt it.

mruberry replied:
I think you're right. In that case let's put the warning in and not worry about adding a new op. We can see how frequent this warning is or if the community complains.

mruberry commented:
We should add a simple argument for updating this to the note. How does the proposed behavior compare to JAX, NumPy, and the Python Array API? Also, we should include the alternative for people who want to keep the existing behavior in the note and consider if we should provide a simpler mechanism to access it.

mruberry commented:
It's difficult to tell how frequently .T is used, but it does appear to be used within Facebook. If this deprecation warning is added we should be able to determine how often it's used on tensors with ndim != 2, too.

Lezcano commented:
Re. Current behaviour.
The proposed behaviour implements the behaviour from the array API, and it deprecates the behaviour from Numpy. This behaviour has stayed in NumPy due to BC, but it's widely agreed that it's certainly not what the user expects. See https://github.com/data-apis/array-api/issues/228 and the links in that issue for further points.

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc mruberry rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc mruberry rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc mruberry rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc mruberry rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc mruberry rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc mruberry rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc mruberry rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

[ghstack-poisoned]

lezcano committed:
Update on "Deprecate x.T on tensors of dimension other than 0 or 2"


**BC-breaking note:**

This PR deprecates the `Tensor.T` are not matrices. An upgrade guide is added to the
documentation for `Tensor.T`.

This PR DOES NOT make this attribute to throw an error when called on a tensor of `dim != 2`,
but this will be its behavior in a future PyTorch release.

cc mruberry rgommers pmeier asmeurer leofang @AnirudhDagar asi1024 emcastillo kmaehashi heitorschueroff

[ghstack-poisoned]

mruberry reviewed (approved):
LGTM!

@anjali411 you're shepherding this stack, right?

(We should be sure that none of our internal code triggers this warning when we land this PR. A brief code search suggests we don't use this internally anywhere, but I might have missed something.)
